Fix Missing Custom Sequences

Malkierian opened the pull request:
Turns out there are sounds loaded into `AudioCollection` that aren't sequences, and the custom sequence loading code was assuming this wasn't the case. This modifies the custom sequence loading code to bypass these known vanilla items that are in `AudioCollection` blocking `seqNum` assignment.

Fixes #2566 but will change indexes assigned with Audio Manager while this problem existed.

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh-linux-compatibility.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/627149031.zip)
  - [soh-linux-performance.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/627149032.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/627149033.zip)
  - [soh-switch.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/627149034.zip)
  - [soh-wiiu.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/627149035.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/627149037.zip)
<!--- section:artifacts:end -->

Malkierian committed:
Modifies custom sequence loading to bypass other sounds being loaded in AudioCollection to fix missing custom sequences on load.

briaguya0 commented on soh/src/code/audio_load.c:
```diff
@@ -1366,16 +1368,23 @@ void AudioLoad_Init(void* heap, size_t heapSize) {
+
     for (size_t i = startingSeqNum; i < startingSeqNum + customSeqListSize; i++) {
+        // ensure that what would be the next sequence number is actually unassigned in AudioCollection
+        while (AudioCollection_GetSequenceName(seqNum) != NULL) {
```
since we're not actually using the sequence name here at all, it probably makes sense to make a new method in `AudioCollection` for checking if a sequence with that key exists in the collection

it should be easy enough to just wrap std::map contains (or std::map count == 0 if cpp 20 ends up being an issue) 
https://en.cppreference.com/w/cpp/container/map/contains
https://en.cppreference.com/w/cpp/container/map/count

Malkierian replied:
Yeah, I wasn't sure adding a new function for just one check would be worth it, but it was an easy enough add.

briaguya0 commented on soh/src/code/audio_load.c:
```diff
@@ -1366,16 +1368,23 @@ void AudioLoad_Init(void* heap, size_t heapSize) {
     int startingSeqNum = MAX_AUTHENTIC_SEQID; // 109 is the highest vanilla sequence
     qsort(customSeqList, customSeqListSize, sizeof(char*), strcmp_sort);
 
+    int seqNum = startingSeqNum;
+
     for (size_t i = startingSeqNum; i < startingSeqNum + customSeqListSize; i++) {
+        // ensure that what would be the next sequence number is actually unassigned in AudioCollection
+        while (AudioCollection_GetSequenceName(seqNum) != NULL) {
+            seqNum++;
+        }
```
it'd be good to add some clarifying comments explaining *why* we need to check this here, and why it's a better solution than just setting `MAX_AUTHENTIC_SEQID` to be a bigger number

Malkierian replied:
Is the fact that there are currently like 20 unused seqNums between the end of the end of the vanilla sequences and the sound fonts at 130-136 causing the main issue enough?

briaguya0 replied:
i think the bigger thing is what you mentioned before about some sequences
> in the low 2000, mid 6000, and 10-12k ranges, too

if it was just 20 unused ones we would probably be fine just setting the max to after the sound fonts

Malkierian replied:
Oh, OK. I'll get some specifics about what those others are and add them to the explanation.

Malkierian replied:
Oh, and I was wrong, they're instruments there, not sound fonts. Those are later I think XD

briaguya0 commented on soh/src/code/audio_load.c:
```diff
@@ -1366,16 +1368,23 @@ void AudioLoad_Init(void* heap, size_t heapSize) {
     int startingSeqNum = MAX_AUTHENTIC_SEQID; // 109 is the highest vanilla sequence
```
is `MAX_AUTHENTIC_SEQID` a lie? are there authentic sequences with higher ids than 109?

Malkierian replied:
No, the sequences end there, but `AudioCollection` has more than just sequences in it, like the sound fonts at 130-136 which were the original issue.

Malkierian replied:
It's a bit cumbersome to have a sequence map in `audio_load.c` and `AudioCollection`, but I'm not sure how to unify them.

Malkierian replied:
So yeah, as mentioned elsewhere, it's more than just the instruments in there. the sound effects are added in as well later on.

briaguya0 commented on soh/src/code/audio_load.c:
```diff
@@ -1346,7 +1346,9 @@ void AudioLoad_Init(void* heap, size_t heapSize) {
     char** seqList = ResourceMgr_ListFiles("audio/sequences*", &seqListSize);
     char** customSeqList = ResourceMgr_ListFiles("custom/music/*", &customSeqListSize);
-    sequenceMapSize = (size_t)(seqListSize + customSeqListSize);
+    // 20 is arbitrary and meant as only a temporary fix. this just gives headroom for the sequences in AudioCollection that take up early indexes that
```
we should probably figure out what the actual number is, and then `#define` it and give it a nice name

Malkierian replied:
There were six within the range of DJ's custom sequence count, but there were also other ones in the low 2000, mid 6000, and 10-12k ranges, too. I really don't know what is going on in `AudioCollection` with all of those non-sequence entries.

Malkierian replied:
I did figure out that I could do an allocation for `audio_load`'s sequenceMap based on the size of `AudioCollection`'s sequenceMap at that point in the load since the other non-sequence audio assets are already loaded into `AudioCollection`, so that can get rid of this part, at least.

briaguya0 reviewed (commented):
Overall I'm super happy to see a fix for this! Left a couple suggestions and a few questions.

Malkierian committed:
Modified `audio_load`'s sequenceMap allocation to utilize `AudioCollection`'s sequenceMap size to account for all audio assets already loaded into that sequenceMap. This gives a non-arbitrary number in addition to the vanilla sequence count to allocate with for `audio_load`'s sequenceMap.

Added `HasSequenceNum` to `AudioCollection` as well to streamline the check against `AudioCollection`'s sequenceMap to skip the non-sequence assets in there.

Added clarification comment for seqNum and MAX_AUTHENTIC_SEQID section.

Malkierian committed:
Clarified comment about AudioCollection seqNum and MAX_AUTHENTIC_SEQID.

Malkierian committed:
A bit more on comment from the last commit.

briaguya0 reviewed (approved):
Beautiful! :shipit:
